prusti-viper/src/encoder/mir_encoder/mod.rs

```diff
@@ -205,15 +285,16 @@ pub trait PlaceEncoder<'v, 'tcx: 'v> {
 
     fn encode_deref(
         &self,
-        encoded_base: vir::Expr,
+        encoded_base: PlaceEncoding,
         base_ty: ty::Ty<'tcx>,
-    ) -> EncodingResult<(vir::Expr, ty::Ty<'tcx>, Option<usize>)> {
+    ) -> EncodingResult<(PlaceEncoding, ty::Ty<'tcx>, Option<usize>)> {
```

Comment:
I find a bit odd to require a `PlaceEncoding` and then immediately call `try_into_expr`. Isn't it better to keep `vir::Expr`? All calls to this function seem to be `.encode_deref(PlaceEncoding::Expr(..), ..)`.

Reply:
well, it's called from encode_projection (because `Deref` is a `ProjectionElem` variant), so it seemed to make sense to be able to handle `PlaceEncoding`s there as well, so for
```
fn deref(x: [&mut u32; 2]) {
    *x[0] = 42;
}
```
we would have that combination.

on the other hand, the implementation is not there yet. i think in general there needs to be more handling in some of the places that `try_into_expr` is currently used to be able to support the full set of nesting possibilities.

Reply:
I see, I missed the call from `encode_projection`. For the following case, however, wouldn't Prusti generate an internal error instead of an unsupported error because of the `try_into_expr`?
```
fn deref(x: [&mut u32; 2]) {
    *x[0] = 42;
}
```
It might be easier to keep the signature of `encode_deref` as it was and do the conversion from `PlaceEncoding` on the call site. That way, you can generate an "unsupported error" in `encode_projection`.

Reply:
i'll rework this then, so that `encode_place` and `encode_projection` are the only two that return `PlaceEncoding`, and mixed/complicated cases like `encode_deref` will return `EncodingError::unsupported` for now